adafruit_led_animation/animation.py

```diff
@@ -0,0 +1,507 @@
+
+
+class Animation:
+    """
+    Base class for animations.
+    """
+    # pylint: disable=too-many-arguments
+    def __init__(self, pixel_object, speed, color, peers=None, paused=False):
+        self.pixel_object = pixel_object
+        self._speed_ns = 0
+        self.speed = speed  # sets _speed_ns
+        self._color = color
+        self._next_update = monotonic_ns()
+        self.pixel_object.auto_write = False
+        self.color = color
+        self.peers = peers if peers else []
+        self._paused = paused
+        self._time_left_at_pause = 0
+
+    def animate(self):
+        """
+        Call animate() from your code's main loop.  It will draw the animation draw() at intervals
+        configured by the speed property (set from init).
+
+        :return: True if the animation draw cycle was triggered, otherwise False.
+        """
+        if self._paused:
+            return False
+
+        now = monotonic_ns()
+        if now < self._next_update:
+            return False
+
+        self.draw()
+        if self.peers:
+            for peer in self.peers:
+                peer.draw()
+        self.show()
+        for peer in self.peers:
+            peer.show()
+
+        self._next_update = now + self._speed_ns
+        return True
+
+    def draw(self):
+        """
+        Animation subclasses must implement draw() to render the animation sequence.
+        """
+
+    def show(self):
+        """
+        Displays the updated pixels.  Called during animates with changes.
+        """
+        self.pixel_object.show()
+
+    def freeze(self):
+        """
+        Stops the animation until resumed.
+        """
+        self._paused = True
+        self._time_left_at_pause = max(0, monotonic_ns() - self._next_update)
+
+    def resume(self):
+        """
+        Resumes the animation.
+        """
+        self._next_update = monotonic_ns() + self._time_left_at_pause
+        self._time_left_at_pause = 0
+        self._paused = False
+
+    def fill(self, color):
+        """
+        Fills the pixel object with a color.
+        """
+        self.pixel_object.fill(color)
+        self.show()
+
+    @property
+    def color(self):
+        """
+        The current color.
+        """
+        return self._color
+
+    @color.setter
+    def color(self, color):
+        if isinstance(color, int):
+            color = (color >> 16 & 0xff, color >> 8 & 0xff, color & 0xff)
+        self._color = color
+        self._recompute_color(color)
+
+    @property
+    def speed(self):
+        """
+        The animation speed in fractional seconds.
+        """
+        return self._speed_ns / 1000000000
+
+    @speed.setter
+    def speed(self, seconds):
+        self._speed_ns = int(seconds * 1000000000)
+
+    def _recompute_color(self, color):
+        pass
+
+    def change_color(self, color):
+        """
+        By default changing color uses the color property.
+        """
```

Comment:
Why is this needed? Why not just use the `color` property setter?

Reply:
It's difficult to override a property setter.  I'm trying to remember why I needed it.  I think _recompute_color is what needs overriding and I forgot I put it there.  Will try to squash this.

Reply:
Oh, now I remember - it ended up there because I didn't use a property on an animation sequence, time to fix that.